### PR TITLE
ci: bootstrap the test env and install the operator

### PR DESCRIPTION
**Pull request by wainersm**
Adds ansible playbooks to setup the test environment. Kubernetes will be
deployed with kubeadm.

The cluster/up.sh is used to bring up the cluster. Afterwards run
build_install_operator.sh to build and install the operator and
ccruntime. Notice it pushes the operator's controller image to a docker
registry running locally so we avoid pushing/pulling to/from quay.io.

I tested in a Fedora 34 x86_64 machine in a Vagrant VM (see the
Vagrantfile added in this commit). If you wish to run likewise:

$ cd tests/e2e
$ vagrant up tests-e2e-ubuntu2004

Signed-off-by: Wainer dos Santos Moschetta <wainersm@redhat.com>

### CODE REVIEW

**Comment by wainersm** [comment]
Hi @jensfr @bpradipt @cmeadors ! I am sending an early version of my scripts for the operator CI with hope that you could give feedbacks. What I am doing wrong?

I tested locally with Vagrant as I explained in the commit message. It apparently installed everything correctly (including the operator and ccruntime). Now I am struggling to get some simple  kata CC tests running so certify it really installed things fine.

Cc @fidencio @stevenhorsman @fitzthum

File: tests/e2e/cluster/up.sh
Comment (fitzthum) [inline tests/e2e/cluster/up.sh:49]: I have been using the role `node-role.kubernetes.io/control-plane-` and/or `node-role.kubernetes.io/master-` depending on Kubernetes version.

File: tests/e2e/cluster/up.sh
Comment (fitzthum) [inline tests/e2e/cluster/up.sh:16]: I've always used the `-pod-network-cidr` and `--cri-socket` args, which don't appear to be in your config file.

**Review by fitzthum — COMMENTED** [review]
A few random comments. I can really only compare this to how I've been setting up my clusters. So if anything my comments are probably just things I've been doing wrong 🙃

Lmk if you still have issues and I can try running it locally. What exactly goes wrong?

**Comment by wainersm** [comment]
Updated this PR to fix a couple of issues, being the most critical the one I commented here: https://github.com/confidential-containers/operator/pull/63/commits/d122881d161f9fd2fb8d518d68ff40f6a85c5a48#diff-09c8e67602614751bc112af4f37558b2a85d24f7492358762de1794d10d36f0aR12

Also bumped Kubernetes to v1.24.0.

I got few kata cc tests passing, and I am debugging others. I am tracking the PRs with fixes to the kata's tests in https://github.com/kata-containers/tests/issues/5077

File: tests/e2e/ansible/install_kubeadm.yml
Comment (wainersm) [reply tests/e2e/ansible/install_kubeadm.yml:14]: Kata CI use that rule: cgroupfs to containerd and systemd for crio. The default of containerd seems to be cgroupfs. Probably we will stick with that rule.

File: tests/e2e/ansible/install_kubeadm.yml
Comment (wainersm) [reply tests/e2e/ansible/install_kubeadm.yml:34]: Don't know, it was on the installation guide with kubeadm :)

Anyway, I have used it to debug some stuffs so it will be useful.

File: tests/e2e/cluster/up.sh
Comment (wainersm) [reply tests/e2e/cluster/up.sh:49]: yeah, that piece is broken, I have manually untainted both...I will need to fix it. Thanks!

File: tests/e2e/cluster/up.sh
Comment (wainersm) [reply tests/e2e/cluster/up.sh:16]: Those options are passed via configuration file: https://github.com/confidential-containers/operator/pull/63/files#diff-f503b1c550a96bc32769671319e42cf289e6a2bcc203d5d16d0fd00b623b8f85

But now that you mentioned, I will double-check those configurations are really taken effect.

File: tests/e2e/ansible/install_kubeadm.yml
Comment (wainersm) [reply tests/e2e/ansible/install_kubeadm.yml]: Bumped to v1.24.0. It seems to be working fine, apart from some kata tests failing which I am debugging still.

**Comment by wainersm** [comment]
hey @fitzthum thanks for your review!

**Comment by wainersm** [comment]
Updated this PR with the machinery to run the kata containers CC tests. 

4 out of 7 tests are passing in my Vagrant VM:
```
IMPORTANT: Assume the image was built with SKOPEO=yes
Cloning into '/tmp/tmp.CSKSgDkgY4/src/github.com/kata-containers/tests'...
remote: Enumerating objects: 26464, done.
remote: Counting objects: 100% (392/392), done.
remote: Compressing objects: 100% (176/176), done.
remote: Total 26464 (delta 228), reused 346 (delta 209), pack-reused 26072
Receiving objects: 100% (26464/26464), 12.15 MiB | 279.00 KiB/s, done.
Resolving deltas: 100% (16782/16782), done.
remote: Enumerating objects: 32, done.
remote: Counting objects: 100% (32/32), done.
remote: Compressing objects: 100% (22/22), done.
remote: Total 32 (delta 14), reused 19 (delta 10), pack-reused 0
Unpacking objects: 100% (32/32), 21.27 KiB | 680.00 KiB/s, done.
From https://github.com/wainersm/kc-tests
 * branch              operator_ci -> FETCH_HEAD
 * [new branch]        operator_ci -> wainersm/operator_ci
Switched to a new branch 'operator_ci'
1..4
ok 1 [cc][agent][kubernetes][containerd] Test can pull an unencrypted image inside the guest
ok 2 [cc][agent][kubernetes][containerd] Test can pull a unencrypted signed image from a protected registry
ok 3 [cc][agent][kubernetes][containerd] Test can pull an unencrypted unsigned image from an unprotected registry
ok 4 [cc][agent][kubernetes][containerd] Test cannot pull an encrypted image inside the guest without decryption key
```

The failing tests are:
```
```

File: tests/e2e/ansible/group_vars/all
Comment (bpradipt) [inline tests/e2e/ansible/group_vars/all:7]: The versions should be aligned to the ones used for building the code
go - 1.18
operator-sdk - v1.23

I think we can have have a VERSIONS file or similar to document the versions in one place.

**Comment by wainersm** [comment]
@bpradipt @jensfr @fitzthum @stevenhorsman @fidencio this PR is ready for another round of reviews.

I added in many places verifications to ensure resources are running before moving to next operation.

It works on the Vagrant VM on my local machine. I also set a job on Jenkins without TEE and it passed there as well (see http://jenkins.katacontainers.io/job/wainer-confidential-containers-operator-main-ubuntu-20.04-x86_64-containerd_qemu_nontee-PR/5/).

@bpradipt suggested to update golang and operator sdk. I will give it a try soon.

Next: finish the Github <-> Jenkins setup so the job will be triggered on pull requests.

**Comment by jensfr** [comment]
@wainersm [this](https://github.com/confidential-containers/operator/pull/70) should fix the error when calling `make undeploy uninstall`

File: tests/e2e/operator.sh
Comment (jensfr) [inline tests/e2e/operator.sh:43]: make install deploy is what developers would use and I think it is good for now. But eventually end users will install the operator in a slightly different way, i.e. from operatorhub.io using OLM (operator lifecycle manager). This is a command I  use to test this and which is close to what the end user will do. 

make IMAGE_TAG_BASE=quay.io/jensfr/confidentials-containers-operator IMG=quay.io/jensfr/confidential-containers-operator:test docker-build docker-push bundle bundle-build bundle-push catalog-build catalog-push && kubectl create -f deploy/deploy_olm.yaml

Note that deploy_olm.yaml does not exist in the git repo yet. I'll send it as a PR soon. This comment is just a heads up and shouldn't block the PR from being merged.

File: tests/e2e/tests_runner.sh
Comment (jensfr) [inline tests/e2e/tests_runner.sh:38]: should this todo be fixed before the merge? does it depend on another pr?

**Comment by stevenhorsman** [comment]
> > Hey Wainer, this is looking like a great start. Just as a data point I tried running this, without vagrant, on a fresh Ubuntu 20.04 VM that I installed Ansible onto. I had a couple of issues which I think were due to not all the waits being in place, so things ran before the operator was properly ready, but after manually triggering the individual steps I got the current 4 tests working. Let me know if there is anything I can do to help and I'll sign up now to re-run the tests once this is ready for review. Thanks!
> 
> Hi @stevenhorsman , thanks for testing it!
> 
> Today my forces has been dragged by trying to put the "waits" on right place and, moreover, make them work (e.g. the flannel checking wasn't). 

I've tried it again and the wait problems seem to be resolved now 🥳 . Specifically (for anyone's reference, including mine late) I created a new Ubuntu 20.04 and as `root` I:
- Installed Ansible with `sudo apt-get -y update && sudo apt-get -y install ansible`
- Checked out Wainer's operator-ci branch
- `cd tests/e2e/`
- `./run-local.sh`

This resulted in the four expected tests passing:
```
ok 1 [cc][agent][kubernetes][containerd] Test can pull an unencrypted image inside the guest
ok 2 [cc][agent][kubernetes][containerd] Test can pull a unencrypted signed image from a protected registry
ok 3 [cc][agent][kubernetes][containerd] Test can pull an unencrypted unsigned image from an unprotected registry
ok 4 [cc][agent][kubernetes][containerd] Test cannot pull an encrypted image inside the guest without decryption key
```

File: tests/e2e/tests_runner.sh
Comment (wainersm) [reply tests/e2e/tests_runner.sh:43]: hi @jensfr ! thanks for the review!

Yes, I'd like to have everything fixed on kata's CCv0 branch before merge this PR. The good news is that there is only this https://github.com/kata-containers/tests/pull/5081 fix pending; once merged I will be able to switch out of my tree.

**Comment by fitzthum** [comment]
I've run this locally using the method that @stevenhorsman describes and I get the same results. I will start looking into adding SEV support.

**Comment by wainersm** [comment]
> I've run this locally using the method that @stevenhorsman describes and I get the same results. I will start looking into adding SEV support.

cool! thx @fitzthum !

**Comment by wainersm** [comment]
I updated the PR where:
 * rebased
 * update Golang to 1.18.5 and operator SDK v1.23
 * added copyrights to most of the files
The tests are still passing: http://jenkins.katacontainers.io/job/wainer-confidential-containers-operator-main-ubuntu-20.04-x86_64-containerd_qemu_nontee-PR/6

**Comment by wainersm** [comment]
> So it doesn't look like there is support for uninstalling anything via the `run-local` script. Is there any plan to add this? It might be important for testing on baremetal.

Good point @fitzthum . The other day I talked with Fabiano about that. I even have some snippets of code for 'undo' the changes; my idea is to put those as tagged blocks on Ansible. I will send what I have so far.

**Comment by wainersm** [comment]
> > So it doesn't look like there is support for uninstalling anything via the `run-local` script. Is there any plan to add this? It might be important for testing on baremetal.
> 
> Good point @fitzthum . The other day I talked with Fabiano about that. I even have some snippets of code for 'undo' the changes; my idea is to put those as tagged blocks on Ansible. I will send what I have so far.

@fitzthum I just pushed the commit https://github.com/confidential-containers/operator/pull/63/commits/e440aed5f2507fb5ecfdbea6ba0d37777d1abdad that implement the "undo" operations on the exit of `run-local` script. I tested the code in parts, tomorrow (today) I should be testing altogether.

File: tests/e2e/operator.sh
Comment (bpradipt) [inline tests/e2e/operator.sh:105]: @wainersm, can you please try adding the following `ignore-not-found=true`  to both `uninstall` and `undeploy` target. 
eg `make uninstall ignore-not-found=true`

This will help avoid making the change recommended in `https://github.com/confidential-containers/operator/pull/70` and keep the Makefile aligned to operator-sdk defaults.

If it doesn't work, then we resort to the change mentioned in PR 70.
cc @jensfr

**Comment by wainersm** [comment]
> > > So it doesn't look like there is support for uninstalling anything via the `run-local` script. Is there any plan to add this? It might be important for testing on baremetal.
> > 
> > 
> > Good point @fitzthum . The other day I talked with Fabiano about that. I even have some snippets of code for 'undo' the changes; my idea is to put those as tagged blocks on Ansible. I will send what I have so far.
> 
> @fitzthum I just pushed the commit [e440aed](https://github.com/confidential-containers/operator/commit/e440aed5f2507fb5ecfdbea6ba0d37777d1abdad) that implement the "undo" operations on the exit of `run-local` script. I tested the code in parts, tomorrow (today) I should be testing altogether.

I created a VM, ran `./run-local.sh -u` and it reverted the changes. Then I connected on it, re-ran `./run-local.sh` then tests all passed. So it there isn't any obvious bug at the moment.

File: tests/e2e/operator.sh
Comment (jensfr) [reply tests/e2e/operator.sh:105]: as discussed in #70 I think adding `ignore-not-found=true`  is the correct change to make. Thanks @bpradipt !

**Comment by fitzthum** [comment]
After running `run-local.sh -u`, `/etc/systemd/system/containerd.service.d/containerd-for-cc-override.conf` is still present, which causes future installations of the runtime to fail with the error `cp: cannot create regular file '/opt/confidential-containers/bin/containerd': Text file busy` coming from the `cc-operator-pre-install-daemon` pod.

It seems like this should be removed [here](https://github.com/fidencio/container-engine-for-cc-payload/blob/main/scripts/container-engine-for-cc-deploy.sh#L43), but I guess it isn't. It seems like the `uninstall_operator` function in `operator.sh` doesn't do anything to uninstall the ccruntime. Maybe this is why we are stuck with the CC containerd?  fwiw I have also run into this issue many times when manually so maybe it's a more general issue?

@wainersm  @bpradipt

File: tests/e2e/operator.sh
Comment (wainersm) [reply tests/e2e/operator.sh]: @bpradipt @jensfr implement a test case for the uninstall operation in https://github.com/confidential-containers/operator/pull/63/commits/b535a90e55745c1afa194dd0ea4c83d91ebabf86 . I added the `ignore-not-found=true` to the Make calls as you suggested.

**Comment by wainersm** [comment]
test

File: tests/e2e/tests_runner.sh
Comment (wainersm) [reply tests/e2e/tests_runner.sh]: done. On last update I removed that TODO, i.e., we are runnning the tests on CCv0 branch, not on my tree.

**Comment by bpradipt** [comment]
> After running `run-local.sh -u`, `/etc/systemd/system/containerd.service.d/containerd-for-cc-override.conf` is still present, which causes future installations of the runtime to fail with the error `cp: cannot create regular file '/opt/confidential-containers/bin/containerd': Text file busy` coming from the `cc-operator-pre-install-daemon` pod.


@fitzthum, I will check this..

**Comment by wainersm** [comment]
@bpradipt @jensfr @fitzthum hello my friends! As you can see on "All checks have passed" widget I am able to trigger the job ("tests-e2e-ubuntu-20.04-x86_64-containerd_kata-qemu") on Jenkins to test this pull request using the CI itself added.

The scripts introduced here can (and will) be improved, but IMO there are already a solid base to start testing the upcoming pull requests for this repository. @fidencio has opened  https://github.com/confidential-containers/operator/pull/72 which I'd like to run through that CI; also @stevenhorsman should be sending something soon to update the payload. Therefore, I am advocating to have this PR approved and merged still today. WDYT?

**Review by bpradipt — APPROVED** [review]
/lgtm
Thanks @wainersm

**Comment by bpradipt** [comment]
> @bpradipt @jensfr @fitzthum hello my friends! As you can see on "All checks have passed" widget I am able to trigger the job ("tests-e2e-ubuntu-20.04-x86_64-containerd_kata-qemu") on Jenkins to test this pull request using the CI itself added.
> 
> The scripts introduced here can (and will) be improved, but IMO there are already a solid base to start testing the upcoming pull requests for this repository. @fidencio has opened #72 which I'd like to run through that CI; also @stevenhorsman should be sending something soon to update the payload. Therefore, I am advocating to have this PR approved and merged still today. WDYT?

Oops, I merged #72 first :-(

**Comment by fidencio** [comment]
> > @bpradipt @jensfr @fitzthum hello my friends! As you can see on "All checks have passed" widget I am able to trigger the job ("tests-e2e-ubuntu-20.04-x86_64-containerd_kata-qemu") on Jenkins to test this pull request using the CI itself added.
> > The scripts introduced here can (and will) be improved, but IMO there are already a solid base to start testing the upcoming pull requests for this repository. @fidencio has opened #72 which I'd like to run through that CI; also @stevenhorsman should be sending something soon to update the payload. Therefore, I am advocating to have this PR approved and merged still today. WDYT?
> 
> Oops, I merged #72 first :-(

That's fine, we can test with #73. :-)

File: tests/e2e/ansible/install_kubeadm.yml
Comment (fitzthum) [inline tests/e2e/ansible/install_kubeadm.yml:82]: Should we do `daemon-reload` here? I've run into some issues with systemd not being able to find the kubelet service after this step (after having uninstalled the CI previously).

**Review by fitzthum — COMMENTED** [review]
Seems fine to merge. We might need a few tweaks in the next week or two. See my one new comment. Wait it's already merged nvm.

**Pull request merged**
